xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/internal/buffer"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+
+// clusterHandler will be given a name representing a cluster. It will then update the CDS policy
+// constantly with a list of Clusters to pass down to XdsClusterResolverLoadBalancingPolicyConfig
+// in a stream like fashion.
+type clusterHandler struct {
+	// A mutex to protect entire tree of clusters.
+	clusterMutex sync.Mutex
+	root         *clusterNode
+
+	// A way to ping CDS Balanccer about any updates or errors to a Node in the tree.
+	// This will either get called from this handler constructing an update or from a child with an error.
+	updateChannel *buffer.Unbounded
+
+	xdsClient      xdsClientInterface
+}
+
+func (ch *clusterHandler) updateRootCluster(clusterName string) /*Return value - the expectation for a stream back to the CDS Policy*/ {
```

Comment:
This needs to take the lock, because it reads and writes `root`.

Reply:
Added a lock/unlock around this function.